### DISCUSSION TITLE
DataFetcherWrapper: Inconsistent description in Directive-Section

### DISCUSSION THREAD

**Discussion by knotenpunkt** [graphql-java/graphql-java#2278] [Q&A] [answered]
Hey

i opend that issue some time ago:

https://github.com/graphql-java/graphql-java-page/issues/45

There you use in the second example a DataFetcherWrapper

Why i must use this Wrapper, why aren't in other cases CompletionStages not treatet correctly?

Is that a good idea all over i use a second, third,...... DataFetcher to wrap it with the help of https://github.com/graphql-java/graphql-java/blob/master/src/main/java/graphql/schema/DataFetcherFactories.java ?

**Comment by knotenpunkt** [comment, marked as answer]
a moment, the wrapper is only there if i need the calculatet result in  between? if the DataFetcherWrapper doesnt need the value in between, i dont have to use the special DataFetcherWrapper, correct?

But then the comment here in your example on the website is still a little bit confusing:

> 
> //
>             // DataFetcherFactories.wrapDataFetcher is a helper to wrap data fetchers so that CompletionStage is handled correctly
>             // along with POJOs
>             //

or are my thinkings here wrong? Do i always need that DataFetcherWrapper in my visitors and directive-Wirinings?

**Reply by knotenpunkt** [reply]
why completions-stage along with Pojos that's not really clear

**Reply by bbakerman** [reply]
The idea behind wrapping a data fetcher is to allow you to call some original data fetcher and then do some work on the results of that call.

`graphql.schema.DataFetcherFactories#wrapDataFetcher` is a helper.  There are 10s of ways to do the same thing (maybe you want to do work BEFORE you call the original data fetcher say)

The code for this is below

All it is doing is handling CompletableFuture (aka Java promises) values be wiring them into to be the actual resolved values of the CF

graphql-java is an async system and hence promises are allowed back as values.

The mapping function in this case will ONLY be called with a resolved value.

The reason for the mix of CFs and POJOs is that sometimes you dont know what the original data fetcher does (is it async / is it immediate values??)  - this code does not care - it handles both outcomes


```

 public static DataFetcher wrapDataFetcher(DataFetcher delegateDataFetcher, BiFunction<DataFetchingEnvironment, Object, Object> mapFunction) {
        return environment -> {
            Object value = delegateDataFetcher.get(environment);
            if (value instanceof CompletionStage) {
                //noinspection unchecked
                return ((CompletionStage<Object>) value).thenApply(v -> mapFunction.apply(environment, v));
            } else {
                return mapFunction.apply(environment, value);
            }
        };
    }
 ```
